src/osmdata-data_frame.cpp

```diff
@@ -0,0 +1,259 @@
+ *                  object of class XmlData and return it in Rcpp::List format,
+ *                  but in this case ignoring the actual geometric data. The
+ *                  returned object contains key-value data only, for return
+ *                  from the R function 'osmdata_data_frame'.
+ *
+ *  Limitations:
+ *
+ *  Dependencies:       none (rapidXML header included in osmdata)
+ *
+ *  Compiler Options:   -std=c++11
+ ***************************************************************************/
+
+#include "osmdata.h"
+
+#include <Rcpp.h>
+
+// Note: This code uses explicit index counters within most loops which use Rcpp
+// objects, because these otherwise require a 
+// static_cast <size_t> (std::distance (...)). This operation copies each
+// instance and can slow the loops down by several orders of magnitude!
+
+/************************************************************************
+ ************************************************************************
+ **                                                                    **
+ **          1. PRIMARY FUNCTIONS TO TRACE WAYS AND RELATIONS          **
+ **                                                                    **
+ ************************************************************************
+ ************************************************************************/
+
+
+//' get_osm_relations
+//'
+//' Return a dual Rcpp::List containing all OSM relations, the firmt element of
+//' which holds `multipolygon` relations, while the second holds all others,
+//' which are stored as `multilinestring` objects.
+//'
+//' @param rels Pointer to the vector of Relation objects
+//' @param nodes Pointer to the vector of node objects
+//' @param ways Pointer to the vector of way objects
+//' @param unique_vals Pointer to a UniqueVals object containing std::sets of all
+//'       unique IDs and keys for each kind of OSM object (nodes, ways, rels).
+//'
+//' @return A dual Rcpp::List, the first of which contains the multipolygon
+//'         relations; the second the multilinestring relations.
+//' 
+//' @noRd 
+Rcpp::DataFrame osm_df::get_osm_relations (const Relations &rels, 
+        const UniqueVals &unique_vals)
+{
+    std::vector <std::string> ids_mp, rel_id_mp; 
+    std::vector <std::string> roles;
+
+    const unsigned int nmp = rels.size ();
+
+    size_t ncol = unique_vals.k_rel.size ();
+    rel_id_mp.reserve (nmp);
+
+    Rcpp::CharacterMatrix kv_mat (Rcpp::Dimension (nmp, ncol));
+    unsigned int count = 0;
+
+    for (auto itr = rels.begin (); itr != rels.end (); ++itr)
+    {
+        Rcpp::checkUserInterrupt ();
+
+        rel_id_mp.push_back (std::to_string (itr->id));
+
+        osm_convert::get_value_mat_rel (itr, unique_vals, kv_mat, count++);
+    }
+
+    Rcpp::DataFrame kv_df;
+    if (rel_id_mp.size () > 0)
+    {
+        kv_mat.attr ("names") = unique_vals.k_rel;
+        kv_mat.attr ("dimnames") = Rcpp::List::create (rel_id_mp, unique_vals.k_rel);
```

Comment:
```suggestion
```
row names are not relevant a can be removed

Reply:
There's no real disadvantage to keeping it, and since R will insert dimnames as soon as the object is returned to the R workspace, I think it's better to keep that (and the following) lines, so at least the dimnames hold useful information.

- *Disadvantage* Only a likely entirely negligible overhead in constructing the list
- *Advantage* Having dimnames that hold useful  information, even if that is entirely redundant.